dotnet/samples/Concepts/Filtering/PIIDetectionWithFilters.cs

```diff
@@ -0,0 +1,404 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.Connectors.OpenAI;
+
+namespace Filtering;
+
+/// <summary>
+/// This example shows how to implement Personal Identifiable Information (PII) detection with Filters using Microsoft Presidio service: https://github.com/microsoft/presidio.
+/// How to run Presidio on Docker locally: https://microsoft.github.io/presidio/installation/#using-docker.
+/// </summary>
+public class PIIDetectionWithFilters(ITestOutputHelper output) : BaseTest(output)
+{
+    /// <summary>
+    /// Use Presidio Text Analyzer to detect PII information in prompt with specified score threshold.
+    /// If the score exceeds the threshold, prompt won't be sent to LLM and custom result will be returned from function.
+    /// Text Analyzer API: https://microsoft.github.io/presidio/api-docs/api-docs.html#tag/Analyzer.
+    /// </summary>
+    [Fact]
+    public async Task PromptAnalyzerAsync()
+    {
+        var builder = Kernel.CreateBuilder();
+
+        // Add Azure OpenAI chat completion service
+        builder.AddAzureOpenAIChatCompletion(
+            TestConfiguration.AzureOpenAI.ChatDeploymentName,
+            TestConfiguration.AzureOpenAI.Endpoint,
+            TestConfiguration.AzureOpenAI.ApiKey);
+
+        // Add logging
+        var logger = this.LoggerFactory.CreateLogger<PIIDetectionWithFilters>();
+        builder.Services.AddSingleton<ILogger>(logger);
+
+        // Add Microsoft Presidio Text Analyzer service and configure HTTP client for it
+        builder.Services.AddHttpClient<PresidioTextAnalyzerService>(client => { client.BaseAddress = new Uri("http://localhost:5001"); });
+
+        // Add prompt filter to analyze rendered prompt for PII before sending it to LLM.
+        // It's possible to change confidence score threshold value from 0 to 1 during testing to see how the logic will behave.
+        builder.Services.AddSingleton<IPromptRenderFilter>(sp => new PromptAnalyzerFilter(
+            sp.GetRequiredService<ILogger>(),
+            sp.GetRequiredService<PresidioTextAnalyzerService>(),
+            scoreThreshold: 0.5));
+
+        var kernel = builder.Build();
+
+        // Example 1: Use prompt with PII
+        var result1 = await kernel.InvokePromptAsync("John Smith drivers license is AC432223");
+        logger.LogInformation("Result: {Result}", result1.ToString());
+
+        /* 
+        Prompt: John Smith drivers license is AC432223
+        Entity type: PERSON. Score: 0.85
+        Entity type: US_DRIVER_LICENSE. Score: 0.6499999999999999
+        Result: Prompt contains PII information. Operation is canceled. 
+        */
```

Comment:
Capturing the data about what went wrong but I'm unsure about returning it as a normal result.

If the goal is to detect PII and then have the operation fail it might be better to throw an exception with the same information. At the moment things fail silently.

Also in a real world sample using a credit card number might be better because it's common to want to send the users name to the LLM (score 0.85) but very unlikely to want to send a credit card number. So it would be useful to show a threshold where user names would pass but credit card numbers would fail.

Reply:
> Capturing the data about what went wrong but I'm unsure about returning it as a normal result. If the goal is to detect PII and then have the operation fail it might be better to throw an exception with the same information. At the moment things fail silently.

Good point, I changed the logic to throw an exception instead. Although I left a comment that it's possible to override `context.Result` to avoid breaking function execution. I believe it still should be useful for automatic function calling scenarios.

> Also in a real world sample using a credit card number might be better because it's common to want to send the users name to the LLM (score 0.85) but very unlikely to want to send a credit card number. So it would be useful to show a threshold where user names would pass but credit card numbers would fail.

Done, I used fake credit card number as an example. The threshold is 0.9 now, and user name with score 0.85 passes while credit card with score 1 fails.